### PR TITLE
REST API notes: support is_todo

### PR DESCRIPTION
**Pull request by charles-e**
fix for this https://github.com/laurent22/joplin/issues/1687

### CODE REVIEW

**Comment by charles-e** [comment]
Happy to write some tests if you would like.

**Comment by tessus** [comment]
Yes, please add some tests as well. Furthermore, please do not fix whitespace errors. It makes the diffs harder to read.

File: ReactNativeClient/lib/services/rest/Api.js [hunk 1]
```diff
@@ -448,10 +448,11 @@ class Api {
 			output.parent_id = folder.id;
 		}
 
-		if ('source_url' in requestNote) output.source_url = requestNote.source_url;
-		if ('author' in requestNote) output.author = requestNote.author;
+		if (requestNote.source_url) output.source_url = requestNote.source_url;
```
Comment (tessus) [inline ReactNativeClient/lib/services/rest/Api.js]: Why did you change `'source_url' in requestNote` to `requestNote.source_url`?

AFAIK in case there is no source_url, there will now be an `undefined` error. I could be wrong though. In any case, I would follow the coding style and not change the style

File: ReactNativeClient/lib/services/rest/Api.js [hunk 2]
```diff
@@ -448,10 +448,11 @@ class Api {
...
+		if (requestNote.author) output.author = requestNote.author;
 		if ('user_updated_time' in requestNote) output.user_updated_time = Database.formatValue(Database.TYPE_INT, requestNote.user_updated_time);
 		if ('user_created_time' in requestNote) output.user_created_time = Database.formatValue(Database.TYPE_INT, requestNote.user_created_time);
+		if (requestNote.is_todo) output.is_todo = requestNote.is_todo;
```
Comment (tessus) [inline ReactNativeClient/lib/services/rest/Api.js]: IMO it should rather be: ` 'is_todo' in requestNote`

Comment (charles-e) [reply ReactNativeClient/lib/services/rest/Api.js]: I am still new to the github workflow.  When I tried to merge there was a conflict with the addition of user-updated/created_time changes so I just added them in with mine.  Likewise the white space changes.

**Comment by charles-e** [comment]
I changed it so that we only set is_todo to (numeric) 1 if is_todo is set to 1 (or ‘1’) in the incoming request.  Test reflects that. 
(or is this too picky?)

File: ReactNativeClient/lib/services/rest/Api.js (hunk 1 above)
Comment (laurent22) [reply ReactNativeClient/lib/services/rest/Api.js]: Yes as @tessus said, please revert this change. If the user provides an empty string, we should set the value to an empty string and not let it undefined.

File: ReactNativeClient/lib/services/rest/Api.js [hunk 3]
```diff
@@ -448,10 +448,13 @@ class Api {
 			output.parent_id = folder.id;
 		}
 
-		if ('source_url' in requestNote) output.source_url = requestNote.source_url;
-		if ('author' in requestNote) output.author = requestNote.author;
+		if (requestNote.source_url) output.source_url = requestNote.source_url;
+		if (requestNote.author) output.author = requestNote.author;
 		if ('user_updated_time' in requestNote) output.user_updated_time = Database.formatValue(Database.TYPE_INT, requestNote.user_updated_time);
 		if ('user_created_time' in requestNote) output.user_created_time = Database.formatValue(Database.TYPE_INT, requestNote.user_created_time);
+		if ('is_todo' in requestNote && requestNote['is_todo'] == 1) {
```
Comment (laurent22) [inline ReactNativeClient/lib/services/rest/Api.js]: No need to check for `requestNote['is_todo'] == 1` - if the user provides a value we should set the property to that value.

File: CliClient/tests/services_rest_Api.js [hunk 4]
```diff
@@ -189,21 +189,62 @@ describe('services_rest_Api', function() {
...
+		expect(response.is_todo).toBeUndefined();
+
+		response = await api.route('POST', 'notes', null, JSON.stringify({
+			title: 'testing 4',
+			parent_id: f.id,
+			is_todo: '1'
+		}));
+		expect(response.title).toBe('testing 4');
+		expect(!!response.id).toBe(true);
+		expect(response.is_todo).toBe(1);
```
Comment (laurent22) [inline CliClient/tests/services_rest_Api.js]: Thanks for adding the tests. To keep it short, please only check the value of "is_todo" (the value of title, id, etc. is already tested in other tests)

**Comment by laurent22** [comment]
Thanks for the PR. Other that the few comments it's looking good.

One important point is that all the changes that add or remove whitespace need to be undone. We want to keep diff focused on only the actual changes. See CONTRIBUTING.md for more info: https://github.com/laurent22/joplin/blob/master/CONTRIBUTING.md#coding-style

**Comment by laurent22** [comment]
Do you have something setup for `git difftool`? In my case I've set it up to vimdiff and so I can run it before committing changes and clean up the code, including removing unnecessary white space changes, console.log statements, etc.

I didn't quite understand your last comment but just so you know, you current branch still have the white space changes. To show why it's an issue, there's about 80 lines that are changed in your branch, but only about 40 of these are relevant. If all commits were like this, with only 50% of the changes doing something, it would make it hard to merge them, make conflicts, and will also make the "blame" tool less useful because the real changes would be lost in the white space changes.

There were talks about adding a linter config to the project to prevent this but it hasn't been done yet.

**Comment by tessus** [comment]
> Whitespace is killing me

I had the same problem, because my editor fixed whitespace errors automatically. I'm using Sublime Text and set the following for the project:

```
    "trim_automatic_white_space": false,
    "trim_trailing_white_space_on_save": false,
    "ensure_newline_at_eof_on_save": false
```

**Comment by charles-e** [comment]
Can someone look again?  I am afraid that messages and/or file views are “crossing in the mail.”  

Just now I can take my two files from laurant22:master and copy them on top of my changes in my branch (which should wipe out my changes vs laurant22:master).  

Then, when I diff on my branch I can see that the only “missing” lines are my 1 line change in api.js and my entire test function in the services_rest_api file.  

Which tells me that my changes don’t have any whitespace differences and we should be good to go. (or I am doing something horribly wrong.)

**Comment by tessus** [comment]
@laurent22 another trick on gh is to add `?w=1` to the URL `https://github.com/laurent22/joplin/pull/1688/files`. in that case no whitspace related changes are shown.

I still don't understand why we don't remove all whitespace errors of the entire codebase in 1 commit. (I know we had that discussion before.)

@charles-e your editor is either removing them automatically, or you have set your git to ignore whitespace warnings, or you have a mess in your branches (which is likely in any case, because of your merge and conflict commits)

**Comment by tessus** [comment]
@charles-e I fixed one of your files now. I will fix the other in a sec.

**Comment by tessus** [comment]
all is fixed now

**Comment by tessus** [comment]
@charles-e this is just great. I fixed all your stuff and now after your last commit it's messed up again. ok I'm done. no more fixes from me - I'm not gonna waste my time

**Comment by laurent22** [comment]
> still don't understand why we don't remove all whitespace errors of the entire codebase in 1 commit. (I know we had that discussion before.)

It needs to be a two steps process:

- first we create a linter config
- then we apply this config to the code base

After that new code needs to use that config every time. Without config new code will just introduce new white space inconsistencies, which will be "auto fixed" again by some editors.

**Comment by hckrgod** [comment]
There is no reason to leave trailing whitespace characters in your project's files, so don't add any.

A git diff --check will tell you if there are any and you should not commit when you see them. So go ahead and switch your editor/IDE to automatically remove them for you.
Below are a few instructions on how to get them removed by your favorite IDE or editor.

Note that except for RubyMine, the following changes will remove trailing white-space on all lines, not only those that you changed.
While this should not be a problem if your project is always clean of trailing spaces, it may not be what you want if you are in the wild west.

In that case you need to take care not to commit trailing spaces. Always git diff --check, kids!
hold on i will tell u more of what i think WE WILL GET THIS FIXED OK
RubyMine

Sent from ProtonMail mobile

-------- Original Message --------

**Comment by tessus** [comment]
> first we create a linter config

JS projects usually use eslint, but there are other tools as well, e.g. [lintspaces](https://www.npmjs.com/package/lintspaces)
Just let me know what you had in mind and I can look into it.

**Comment by tessus** [comment]
@hckrgod you are totally missing the point and unless you have particpated in the full discussion about whitespace errors (in several gh issues, on the forum, and private discussions), please stop commenting on it. my comments were not directed to the public.

**Comment by charles-e** [comment]
Honestly, I am not _trying_ to anger you.  Eventually I think I can be allowed to “run with scissors.”
I have rolled back my previous change and then edited the file online here in github to simply remove title checks in my test (per Laurent’s comment).

**Comment by hckrgod** [comment]
ok this is similiar but not exact maybe it will make u fix this i have been trying too and its a mystery but I just started to find and fix trailing whitespace in your commits. Bypass # it with the --no-verify option to-commit. # detect platform platform="win" uname_result=`uname` if [[ "$uname_result" == "Linux" ]]; then platform="linux" elif [[ "$uname_result" == "Darwin" ]]; then platform="mac" fi # change IFS to ignore filename's space in |for| IFS=" " # remove trailing whitespace in modified lines for line in diff --check --cached | sed '/^[+-]/d'` ; do # get file name if [[ "$platform" == "mac" ]]; then file="`echo $line | sed -E 's/:[0-9]+: .*//'`" line_number="`echo $line | sed -E 's/.*:([0-9]+).*/\1/'`" else file="`echo $line | sed -r 's/:[0-9]+: .*//'`" line_number="`echo $line | sed -r 's/.*:([0-9]+).*/\1/'`" fi # since $file in working directory isn't always equal to $file in index, # we backup it; thereby we can add our whitespace fixes without accidently # adding unstaged changes backup_file="${file}.working_directory_backup" cat "$file" > "$backup_file" -- "$file" # discard unstaged changes in working directory # remove trailing whitespace in $file (modified lines only) if [[ "$platform" == "win" ]]; then # in windows, `sed -i` adds ready-only attribute to $file (I don't kown why), so we use temp file instead sed "${line_number}s/[[:space:]]*$//" "$file" > "${file}.bak" mv -f "${file}.bak" "$file" elif [[ "$platform" == "mac" ]]; then sed -i "" "${line_number}s/[[:space:]]*$//" "$file" else sed -i "${line_number}s/[[:space:]]*$//" "$file" fi git add "$file" # to index, so our whitespace changes will be committed # restore unstaged changes in $file from its working directory backup, fixing # whitespace that we fixed above sed "${line_number}s/[[:space:]]*$//" "$backup_file" > "$file" rm "$backup_file" [[ "$platform" == "mac" ]] || e_option="-e" # mac does not understand -e echo $e_option "Removed trailing whitespace in \033[31m$file\033[0m:$line_number" done echo # If there still are whitespace errors, print the offending file names and fail-index --check --cached $against -- # Now we can commit exit
wow this is baffling me i will get it what do u think did u try what u sais in last msg ??? it didnt work right ???
Sent from ProtonMail mobile

-------- Original Message --------

**Comment by tessus** [comment]
> Honestly, I am not trying to anger you.

I know that. My bluntness came across wrongly again. Sorry about that. I'm not angry at or with you. I write things without emotional undertone (at least that's how it was meant), but it can be perceived as such. I have to work on that.

> remove title checks in my test 

Thank you, LGTM.

**Comment by charles-e** [comment]
Great - I had to look up “LGTM”.

**Comment by laurent22** [comment]
> Great - I had to look up “LGTM”.

An acronym I've also learnt in this project :-)

**Comment by laurent22** [comment]
> JS projects usually use eslint, but there are other tools as well, e.g. [lintspaces](https://www.npmjs.com/package/lintspaces)

I had bad experience with eslint in Sublime Text. It would freeze the editor for a second while running the tool. The setup was also surprisingly complex (needed to install a framework plugin, then another plugin, which wasn't well supported) and it would be hard to make it work on Windows (just gave up trying).

So I guess number 1 priority would be a linter that supports Sublime properly, without freezing the editor, and works on Windows too.

**Comment by charles-e** [comment]
I am using atom on mac right now, but sublime is fine if that will make things easier.

**Comment by laurent22** [comment]
I assume if the linter we pick supports Sublime it will support Atom too. Atom as well as VSCode are probably more popular than Sublime so better supported.

**Comment by tessus** [comment]
> I had bad experience with eslint in Sublime Text

Ahh, ok, I was talking about a much esiar solution: an eslint profile that will be run in travis for every PR that is submitted. If it fails the PR will have a red X. In that case people who submit a PR would then run eslint before submitting the code.

I never would have thought of adding a linter to the editor, because of its complexity. Now I'm starting to understand why you are reluctant doing this in the first place. Yes, with this approach I would have my doubts as well.

**Comment by laurent22** [comment]
I wouldn't want this on CI I think. I don't want things to fail at this stage just because there's an extra space in one file, and I also prefer not to make CI more complex by adding more dependencies, more code, etc. just for this.

Maybe as a pre-commit hook it would make sense because by then the code is not yet in git and can be fixed. That's why I like when it's an editor plugin - the code is fixed as you go, you see what's changed, and by the time you commit (and well before it's processed by CI) you know the code is valid.

**Comment by tessus** [comment]
In that case, we should use `.editorconfig`. AFAIK there are plugins for all major editors / IDEs. Also, you have already added a `joplin.sublime-project` file to the repo. You could also add these 3 settings (and a few more, like use-tabs):

```
    "trim_automatic_white_space": true,
    "trim_trailing_white_space_on_save": true,
    "ensure_newline_at_eof_on_save": true
```

**Comment by laurent22** [comment]
Thanks for the PR @charles-e! Just for info I've removed a few tests because they were already done in "should create note". I realise that one test may work in one place and fail in another, but in this case it's relatively unlikely so we can remove them.

**Comment by charles-e** [comment]
Nice.  This is my first time contributing to anything using github.  Glad I can help.  Of course, this is my own self interest: 
A) Joplin has a good, critical core of functionality that is not reliant on third parties.
B) There is still a ton of low hanging fruit for enhancement (so I can have motivation to learn by fixing.)
